homeassistant/components/roborock/image.py

```diff
@@ -0,0 +1,142 @@
+from vacuum_map_parser_base.config.color import ColorsPalette
+from vacuum_map_parser_base.config.image_config import ImageConfig
+from vacuum_map_parser_base.config.size import Sizes
+from vacuum_map_parser_roborock.map_data_parser import RoborockMapDataParser
+
+from homeassistant.components.image import ImageEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util import slugify
+import homeassistant.util.dt as dt_util
+
+from .const import DOMAIN, IMAGE_CACHE_INTERVAL, IMAGE_DRAWABLES
+from .coordinator import RoborockDataUpdateCoordinator
+from .device import RoborockCoordinatedEntity
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Roborock image platform."""
+
+    coordinators: dict[str, RoborockDataUpdateCoordinator] = hass.data[DOMAIN][
+        config_entry.entry_id
+    ]
+    entities = list(
+        chain.from_iterable(
+            await asyncio.gather(
+                *(create_coordinator_maps(coord) for coord in coordinators.values())
+            )
+        )
+    )
+    async_add_entities(entities)
+
+
+class RoborockMap(RoborockCoordinatedEntity, ImageEntity):
+    """A class to let you visualize the map."""
+
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        unique_id: str,
+        coordinator: RoborockDataUpdateCoordinator,
+        map_flag: int,
+        starting_map: bytes,
+        map_name: str,
+    ) -> None:
+        """Initialize a Roborock map."""
+        RoborockCoordinatedEntity.__init__(self, unique_id, coordinator)
+        ImageEntity.__init__(self, coordinator.hass)
+        self._attr_name = map_name
+        self.parser = RoborockMapDataParser(
+            ColorsPalette(), Sizes(), IMAGE_DRAWABLES, ImageConfig(), []
+        )
+        self._attr_image_last_updated = dt_util.utcnow()
+        self.map_flag = map_flag
+        self.cached_map = self._create_image(starting_map)
+
+    def is_map_valid(self) -> bool:
+        """Update this map if it is the current active map, and the vacuum is cleaning."""
+        return (
+            self.map_flag == self.coordinator.current_map
+            and self.image_last_updated is not None
+            and self.coordinator.roborock_device_info.props.status is not None
+            and self.coordinator.roborock_device_info.props.status.in_cleaning
+        )
+
+    def _handle_coordinator_update(self):
+        # Bump last updated every third time the coordinator runs, so that async_image will be called and
+        # we will evaluate on the new coordinator data if we should update the cache.
+        if (
+            dt_util.utcnow() - self.image_last_updated
+        ).total_seconds() > IMAGE_CACHE_INTERVAL:
```

Comment:
Is it worth skipping updates when `is_map_valid` is False? (no changes because its not cleaning, or map is not active, etc)

Reply:
That makes sense to me - changed.